### ISSUE TITLE
Consider a pattern to override responses after setup

### ISSUE THREAD

**Issue by slightlytyler** [mswjs/msw#128]
I like the look of this library but to be able to use it I'd need the ability to override a response from within a test. Something like:

```
it('should handle some weird edge case response', () => {
  mockedHttpClient.onGet('/endpoint').replyOnce(() => ({
    // ...
  }));
  const { getByText } = render(....);
  // ...
});
```

**Additional context**
Jest handles this very well within their mock functions using `mockImplementationOnce` which I think is a good source of inspiration

https://jestjs.io/docs/en/mock-function-api.html

**Comment by mattrothenberg** [comment]
I definitely agree with @slightlytyler here.

The ergonomics of the `mockImplemenationOnce` API feel spot-on here, where a subject under test needs to be able to mock the behavior of a dependency in an on-demand fashion.

Pseudo-code below
```tsx
import * as React from "react";
import { setupServer } from "msw/node";
import { render, screen } from "@testing-library/react";

import { ConnectionStatusBox } from "components/ConnectionStatusBox";

describe(ConnectionStatusBox, () => {
  const server = setupServer();

  beforeAll(() => {
    server.listen();
  });

  afterAll(() => {
    server.close();
  });

  it("shows a generic error message if a failure occurs when fetching status information", async () => {
    server.get("/api/connection", () => {
      return res(
        ctx.status(500, "You lose!!!"),
        ctx.json({
          message: "An unexpected error occurred",
        })
      );
    })

    render(<ConnectionStatusBox />);
    await screen.findByText(/an unexpected error occurred/gi);
  });

  it("shows the connection status details if we fetch the connection succesfully", async () => {
    server.get("/api/connection", () => {
      return res(
        ctx.status(200),
        ctx.json({
          id: "1234"
        })
      );
    })

    render(<ConnectionStatusBox />);
    await screen.findByText(/status 1234/gi);
  });
});
```

**Comment by kettanaito** [comment]
**What I'd expect from such API:**

1. Be isomorphic for both browser and server. 
1. Request handler override must use the same request handler's API (both to remain consistent, and expose all the regular features).
1. Request path, or operation name, can serve as an identifier to link an override to an existing request handler.
1. Support one-time response override.
1. Support persistent response override (as in until explicitly cancelled).

> Although I see the most use of such override API in NodeJS environment mainly (for unit/integration tests), I think it's more of a technical challenge to find a way to have the same API to affect client-side request handlers as well (for example, as a part of the application's runtime).

**Comment by kettanaito** [comment]
A quick update on this: finished the implementation of this feature for the NodeJS API (`setupServer`), switched to the client-side API for `setupWorker`. Expect this feature soon.

**Comment by kettanaito** [comment]
**Released in [`0.18.0`](https://github.com/mswjs/msw/releases/tag/v0.18.0).**

@slightlytyler, thanks once more for brining this up. Please, give it a try and let me know if you experience any issues.
